gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ApplicationServiceImpl.java

```diff
@@ -540,14 +595,38 @@ private Single<Application> create0(String domain, Application application, User
         application.setCreatedAt(new Date());
         application.setUpdatedAt(application.getCreatedAt());
 
+        var secretSettings = this.applicationSecretConfig.toSecretSettings();
+        application.setSecretSettings(List.of(secretSettings));
+
+        // here the client_secret has been generated, we can generate the HashValue here
+        // as we do not want to store the clear text value for the client Secret
+        // we keep the value and force the value into OAuth setting to null
+        // in order to restore it after the creation
+        var applicationSettings = application.getSettings();
+        final var rawSecret = applicationSettings.getOauth().getClientSecret();
+        if (rawSecret != null) {
+            // PUBLIC client doesn't need to have secret, so wa have to test it before generated the hash
+            applicationSettings.getOauth().setClientSecret(null);
+            var clientSecret = this.clientSecretService.generateClientSecret(rawSecret, secretSettings);
+            application.setSecrets(List.of(clientSecret));
+        }
+
         // check uniqueness
         return checkApplicationUniqueness(domain, application)
                 // validate application metadata
                 .andThen(validateApplicationMetadata(application))
+                .flatMap(this::validateApplicationAuthMethod)
                 // set default certificate
                 .flatMap(this::setDefaultCertificate)
                 // create the application
-                .flatMap(applicationRepository::create)
+                .flatMap(applicationRepository::create).map(app -> {
+                    if (app.getSettings() != null && app.getSettings().getOauth() != null) {
+                        app.getSettings().getOauth().setClientSecret(rawSecret);
+                    } else {
+                        LOGGER.warn("Unable to restore the clientSecret into the application settings, settings are null.");
+                    }
+                    return app;
+                })
                 // create the owner
                 .flatMap(application1 -> {
                     if (principal == null || principal.getAdditionalInformation() == null || StringUtils.isEmpty(principal.getAdditionalInformation().get(Claims.organization))) {
```

Comment:
suggestion: 
StringUtils.isEmpty() is deprecated it can be replaced by !StringUtils.hasText